datafusion/expr/src/type_coercion/binary.rs

```diff
@@ -289,15 +290,169 @@ fn bitwise_coercion(left_type: &DataType, right_type: &DataType) -> Option<DataT
     }
 }
 
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+enum TypeCategory {
+    Array,
+    Boolean,
+    Numeric,
+    // String, well-defined type, but are considered as unknown type.
+    DateTime,
+    Composite,
+    Unknown,
+    NotSupported,
+}
+
+fn data_type_category(data_type: &DataType) -> TypeCategory {
```

Comment:
This could also be something like 

```rust
impl From<&DataType> for TypeCategory { 
...
```

And then you create a `TypeCategory` like
```rust
let category = TypeCategory::from(&type)
```